### DISCUSSION TITLE
Scrolling dashboard tables

### DISCUSSION THREAD

**Discussion by manson** [kamalshkeir/kago#17] [General]
When scrolling table in dashboard the data is being loaded with new rows (next pages) and time fields for newly loaded rows are not displaying (empty), although data exists and when open the form, it is shown ok.

**Comment by kamalshkeir** [comment]
> When scrolling table in dashboard the data is being loaded with new rows (next pages) and time fields for newly loaded rows are not displaying (empty), although data exists and when open the form, it is shown ok.

@manson can you try remove assets folder and make it clone it again ? , i have encountred this problem when i was changing time to unix, i remember i fix it but you should regenerate admin assets

**Reply by kamalshkeir** [reply]
@manson , my bad, i fixed the search last time, fixed now, you can regenerate assets folder

**Reply by manson** [reply]
Probably you have forgotten to push changes? ```Go get -u``` does nothing this time...

**Reply by kamalshkeir** [reply]
> Probably you have forgotten to push changes? ```Go get -u``` does nothing this time...

No i did not change in code, only the assets folder should be updated

**Reply by manson** [reply]
> No i did not change in code, only the assets folder should be updated
Recreated assets, hardreloaded page and gets the same
![image](https://user-images.githubusercontent.com/176285/219953260-c3b20d13-253b-45c1-a082-dc5c592ca227.png)

**Reply by manson** [reply]
first 10 records shows time, next loaded - no. Not a big issue but not perfect :-)

**Reply by kamalshkeir** [reply]
@manson you deleted the assets folder ?

**Reply by manson** [reply]
> @manson you deleted the assets folder ?
yes

        github.com/kamalshkeir/kago v1.2.7
	github.com/kamalshkeir/kenv v1.0.0
	github.com/kamalshkeir/klog v1.0.0
	github.com/kamalshkeir/kmux v1.9.2
	github.com/kamalshkeir/korm v1.7.1
	github.com/kamalshkeir/ksbus v1.1.1

Recreated assets, hardreloaded page and gets the same

**Reply by kamalshkeir** [reply]
> > @manson you deleted the assets folder ?
> > yes
> 
> ```
>     github.com/kamalshkeir/kago v1.2.7
> github.com/kamalshkeir/kenv v1.0.0
> github.com/kamalshkeir/klog v1.0.0
> github.com/kamalshkeir/kmux v1.9.2
> github.com/kamalshkeir/korm v1.7.1
> github.com/kamalshkeir/ksbus v1.1.1
> ```
> 
> Recreated assets, hardreloaded page and gets the same

clear the cache, it should be working

**Reply by kamalshkeir** [reply]
@manson i tested again 

![Sans titre](https://user-images.githubusercontent.com/54605903/219953870-60e84022-4ab9-4c5f-9061-169175029f21.png)

**Reply by manson** [reply]
> clear the cache, it should be working

Damn, I did clear the cache twice it did not help, but now after 3d time it worked... miracle... :-( But do not go away I have another issue for you :-)

**Comment by manson** [comment]
@kamalshkeir 
Another one thing found - when I do many writings to database from different parallel threads, sometimes I see exception "DATABASE BUSY" or something like this. I did not go further to investigate, but can suppose that database works in ```Multi-thread``` mode that although permits multithreading but connection to a database cannot be used from different threads (as it seems to be now), Say alone ```Single-thread``` mode when no any mutexes at all. But definitely this is not a ```Serialized``` mode that drops all restrictions. If this is like that, so the korm should mutex connections or use database in serialized mode...
Here what internet says:"The SQLITE_BUSY error occurs when the database is locked by another process or thread and cannot be accessed. This can happen when multiple threads or processes are trying to access the same database at the same time. To resolve this issue, you should use a database connection pool to manage the database connections or implement a locking or synchronization mechanism to ensure that only one thread or process can access the database at a time. You can also use the SQLite busy timeout setting to configure the amount of time SQLite will wait before it returns a busy error."

**Reply by kamalshkeir** [reply]
> @kamalshkeir Another one thing found - when I do many writings to database from different parallel threads, sometimes I see exception "DATABASE BUSY" or something like this. I did not go further to investigate, but can suppose that database works in `Multi-thread` mode that although permits multithreading but connection to a database cannot be used from different threads (as it seems to be now), Say alone `Single-thread` mode when no any mutexes at all. But definitely this is not a `Serialized` mode that drops all restrictions. If this is like that, so the korm should mutex connections or use database in serialized mode... Here what internet says:"The SQLITE_BUSY error occurs when the database is locked by another process or thread and cannot be accessed. This can happen when multiple threads or processes are trying to access the same database at the same time. To resolve this issue, you should use a database connection pool to manage the database connections or implement a locking or synchronization mechanism to ensure that only one thread or process can access the database at a time. You can also use the SQLite busy timeout setting to configure the amount of time SQLite will wait before it returns a busy error."

i though it was thread safe, because i test it many times from handlers with stress test, so many requests in parallel on the same endpoint that query some data but maybe the cache was preventing this.

can you try to test it with postgres or mysql to see if the behavior is the same ?
because postgres and mysql use a pool of connections, so you should be able to do requests in parallel

**Reply by manson** [reply]
I don't use mysql or postgress here at home, where I do my small project with Kago (and don't want to install'em). Although I have used them before, my latest projects on a work uses MSSQL server. Anyway, I'm sure mysql and postgress will work well in this situation. I don't know how you use db connection in this case, but specifically for SQLITE would suggest the next code by using the pool, already integrated into sqlite:
```
import "github.com/mattn/go-sqlite3"

// Create a database pool with size 5 and a timeout of 10 minutes
dbPool, err := sqlite3.NewDBPool("sqlite3", "mydb.sqlite", 5, 10*time.Minute)
if err != nil {
    // handle error
}

// Get a connection from the pool
conn, err := dbPool.Get()
if err != nil {
    // handle error
}
```
Probably there could be more controll like sql waits time and dynamic pool, but this is another question. I think at least a config in .env for sqlite pool size would be perfect and enough. Say default 5 connections, but user may set for his use case

**Reply by manson** [reply]
``` i test it many times from handlers with stress test, so many requests in parallel on the same endpoint that query some data ``` - in my use case, I just did 1 route, in handler of which I do 5 writings to database in cycle. And calls of this route I even did not by a some kind of stress test tool, but simply typing in console ```curl http://localhost:9313/myroute``` then enter, up button and enter, up button and enter (Yea, I know I'm lazy :-) But definitely later I would stress test my app...

**Reply by kamalshkeir** [reply]
> I don't use mysql or postgress here at home, where I do my small project with Kago (and don't want to install'em). Although I have used them before, my latest projects on a work uses MSSQL server. Anyway, I'm sure mysql and postgress will work well in this situation. I don't know how you use db connection in this case, but specifically for SQLITE would suggest the next code by using the pool, already integrated into sqlite:
> 
> ```
> import "github.com/mattn/go-sqlite3"
> 
> // Create a database pool with size 5 and a timeout of 10 minutes
> dbPool, err := sqlite3.NewDBPool("sqlite3", "mydb.sqlite", 5, 10*time.Minute)
> if err != nil {
>     // handle error
> }
> 
> // Get a connection from the pool
> conn, err := dbPool.Get()
> if err != nil {
>     // handle error
> }
> ```
> 
> Probably there could be more controll like sql waits time and dynamic pool, but this is another question. I think at least a config in .env for sqlite pool size would be perfect and enough. Say default 5 connections, but user may set for his use case

@manson there is no such things, the pool is handled by the standard library database/sql package , and here you can find all of them: https://github.com/kamalshkeir/korm#global-vars

what you can try also, is using different sqlite driver like go-sqlite, and remove sqlitedriver.Use()

and here is an example of database/sql pkg, so my connection when u call GetConnection is the pool:

```go
var pool *sql.DB // Database connection pool.

func main() {
	id := flag.Int64("id", 0, "person ID to find")
	dsn := flag.String("dsn", os.Getenv("DSN"), "connection data source name")
	flag.Parse()

	if len(*dsn) == 0 {
		log.Fatal("missing dsn flag")
	}
	if *id == 0 {
		log.Fatal("missing person ID")
	}
	var err error

	// Opening a driver typically will not attempt to connect to the database.
	pool, err = sql.Open("driver-name", *dsn)
	if err != nil {
		// This will not be a connection error, but a DSN parse error or
		// another initialization error.
		log.Fatal("unable to use data source name", err)
	}
	defer pool.Close()

	pool.SetConnMaxLifetime(0)
	pool.SetMaxIdleConns(3)
	pool.SetMaxOpenConns(3)
```

So you can also use something like `korm.GetConnection().SetConnMaxLifetime()`

**Reply by kamalshkeir** [reply]
> `i test it many times from handlers with stress test, so many requests in parallel on the same endpoint that query some data` - in my use case, I just did 1 route, in handler of which I do 5 writings to database in cycle. And calls of this route I even did not by a some kind of stress test tool, but simply typing in console `curl http://localhost:9313/myroute` then enter, up button and enter, up button and enter (Yea, I know I'm lazy :-) But definitely later I would stress test my app...

yeah so it's multiple write, multiple reads should be ok i think

**Reply by manson** [reply]
```What you can also try is to use another sqlite driver like go-sqlite, and remove sqlitedriver.Use()``` - no, I want to use it in the default way of the framework :-)

```there is no such thing, the pool is handled by the standard database/sql package library, and here you can find them all``` - ok I will play with them

```Yes, so multiple writes, multiple reads should be ok, I think``` - well, this is a common situation and at any time (unexpected user behavior, writing logs or other) you can face it, so it should definitely be handled in advance. It is not a big problem for me - I can use a self-made pool or a kind of other workaround, but all in all I would prefer not to think of it if using a framework. Besides, this is only the sqlite "issue", so if use othe database (I would use mongo later) it won't be an issue.

**Reply by kamalshkeir** [reply]
@manson i played with it, and indeed , this code will return a busy error only when concurrent writing:
```go
package main

import (
	"github.com/kamalshkeir/klog"
	"github.com/kamalshkeir/korm"
        "github.com/kamalshkeir/sqlitedriver"
)

func main() {
        sqlitedriver.Use()
	err := korm.New(korm.SQLITE, "db")
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	srv := korm.WithDashboard()
	korm.WithShell()

	// for i := 0; i < 5; i++ {
	// 	go func(i int) {
	// 		iStr := strconv.Itoa(i)
	// 		users, err := korm.Model[korm.User]().Insert(&korm.User{
	// 			Uuid:     iStr + "-uuid",
	// 			Email:    "email-" + iStr + "@gmail.com",
	// 			Password: "somepass",
	// 		})
	// 		if !klog.CheckError(err) {
	// 			fmt.Println(users)
	// 		}
	// 	}(i)
	// }

	srv.Run("localhost:9313")
}

```
 so i tested github.com/mattn/go-sqlite3, and it work like it should, but the only problem is that mattn go-sqlite3 is not completely written in go like `modernc.org/sqlite` i am using, but i think it is better to replace modernc in this case, what do you think ?
 
 mattn example
 
 ```go
 package main

import (
	"database/sql"

	"github.com/kamalshkeir/klog"
	"github.com/kamalshkeir/korm"
	_ "github.com/mattn/go-sqlite3"
)

func main() {
	db, err := sql.Open("sqlite3", "db.sqlite")
	if klog.CheckError(err) {
		return
	}
	err = korm.NewFromConnection(korm.SQLITE, "db", db)
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	srv := korm.WithDashboard()
	korm.WithShell()

	for i := 0; i < 5; i++ {
		go func(i int) {
			iStr := strconv.Itoa(i)
			users, err := korm.Model[korm.User]().Insert(&korm.User{
				Uuid:     iStr + "-uuid",
				Email:    "email-" + iStr + "@gmail.com",
				Password: "somepass",
			})
			if !klog.CheckError(err) {
				fmt.Println(users)
			}
		}(i)
	}

	srv.Run("localhost:9313")
}

 ```

**Reply by manson** [reply]
```what do you think ?``` - to tell the truth I, as a user of a kago framework, do not care how it will be implemented in the background while it performs as should without issues. If the fact that lib is not written in go will not have side effects (different OSes or else) I think you can change the lib. But to keep framework "clean" in sense of pure Go code (if you have such intend), keep watching if they will improve current lib. And if yes - come back to it. The only thing is this all should not touch the user's code after upgrading of kago.
Or maybe you would like to implement your own code above ```modernc.org/sqlite``` - I don't know, it is your choice :-) But currently, if changing the sqlite lib will not affect user's code now and later, I would prefer it should be replaced to a non-problem lib.
Tests will be needed...

**Reply by kamalshkeir** [reply]
@manson , done, you can update korm to v1.7.2 and sqlitedriver to v1.3.0, when i add it, it break the foreign keys for sqlite, but then i fix it, so this example will work :

```go
import (
	"fmt"
	"strconv"

	"github.com/kamalshkeir/klog"
	"github.com/kamalshkeir/korm"
	"github.com/kamalshkeir/sqlitedriver"
)

type Profile struct {
	Id     uint `korm:"pk"`
	UserId uint `korm:"fk:users.id:cascade:cascade"`
}

func main() {
	sqlitedriver.Use()
	err := korm.New(korm.SQLITE, "db")
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	srv := korm.WithDashboard()
	korm.WithShell()

	err = korm.AutoMigrate[Profile]("profile")
	if klog.CheckError(err) {
		return
	}

	for i := 0; i < 5; i++ {
		go func(i int) {
			iStr := strconv.Itoa(i)
			users, err := korm.Model[korm.User]().Insert(&korm.User{
				Uuid:     iStr + "-uuid",
				Email:    "email-" + iStr + "@gmail.com",
				Password: "somepass",
			})
			if !klog.CheckError(err) {
				fmt.Println(users)
			}
		}(i)
	}

	srv.Run("localhost:9313")
}

```

**Reply by manson** [reply]
Thanks a lot! Unfortunately, will test it later, probably tomorrow - currently away from my computer... But I really glad!

**Reply by manson** [reply]
Made a fast check - looks like it works. Will make more intensive check later. The only strange thing is that database has now sqlite3 extension. Not an issue but first after update I thought it dropped database when it showed nothing :-)

Just playing with SSE and persistent notifications. Currently, implemented this logic: 
1) Put the notification into a memory cache with TTL 10 secs
2) Send it to a user (internal broker uses messages queue with channel)
3) If successful sending - remove from cache
4) If failed sending - move data from cache to the database with a more length TTL (1 minute)
5) If something hangs - record in cache is cleared after it's TTL and moves to database as well
6) When user reconnects, it gets saved and not expired notifications from database, and they are removed from database
7) subscribe to new notifications

Not 100% reliable (I do not use message ID from client but rather success or not success sending) but it suits in my use case...

**Reply by kamalshkeir** [reply]
> Made a fast check - looks like it works. Will make more intensive check later. The only strange thing is that database has now sqlite3 extension. Not an issue but first after update I thought it dropped database when it showed nothing :-)
> 
> Just playing with SSE and persistent notifications. Currently, implemented this logic: 
> 1) Put the notification into a memory cache with TTL 10 secs
> 2) Send it to a user (internal broker uses messages queue with channel)
> 3) If successful sending - remove from cache
> 4) If failed sending - move data from cache to the database with a more length TTL (1 minute)
> 5) If something hangs - record in cache is cleared after it's TTL and moves to database as well
> 6) When user reconnects, it gets saved and not expired notifications from database, and they are removed from database
> 7) subscribe to new notifications
> 
> Not 100% reliable (I do not use message ID from client but rather success or not success sending) but it suits in my use case...
> 

great 👍 , keep me in touch

**Reply by kamalshkeir** [reply]
@manson , kmux no longer support regex, i implemented a radix tree instead of saving routes inside a slice, it is faster : 

```sh
// BenchmarkNew-4                6398119               188.3 ns/op            29 B/op          2 allocs/op
// BenchmarkNewWithParam-4       4636768               272.2 ns/op            60 B/op          2 allocs/op
// BenchmarkNewWith2Param-4      4083430               289.3 ns/op            64 B/op          2 allocs/op
// BenchmarkKmux-4                  1000000              1049 ns/op             464 B/op          7 allocs/op
// BenchmarkKmuxWithParam-4         1000000              1290 ns/op             869 B/op          9 allocs/op
// BenchmarkKmuxWith2Param-4         393536              2565 ns/op             893 B/op          9 allocs/op
 ```

 be aware , there is no more /name:str or others , only '/:name' is allowed, and '/test/*wild' for the wild card route
 to see how i benchmarked this : https://github.com/kamalshkeir/kmux/blob/master/bench/bench_test.go

**Reply by manson** [reply]
@kamalshkeir 
Very well. I think it is really better. I don't like to use regex without big necessity, but ok, you've decided so, I accepted at that time. I know that regex is something slow, especially in some patterns - greedy. So I am very glad, way to go! :-)

**Reply by manson** [reply]
Just came back to a project, so if something face wrong, will respond.

**Reply by manson** [reply]
> be aware , there is no more /name:str or others

You have forgotten to change readme in kmux rep

BTW - where did ```kmux.M{}``` go? I used it...

**Reply by manson** [reply]
> be aware , there is no more /name:str or others , only '/:name' is allowed

Unfortunately, right away from the update I face an exception. If you leave or will have only 1 route like ```sbus.App.POST("/:id", v1.ProcessSomething)``` you will have: 
```github.com/kamalshkeir/kmux.(*node).insertChild[line:579]:wildcard segment ':id' conflicts with existing children in path '/:id'```
Doesn't matter the name ":id" or ":whatever" you will have this exception
```
go: upgraded github.com/kamalshkeir/kmux v1.9.2 => v1.9.4
go: upgraded github.com/kamalshkeir/korm v1.7.2 => v1.7.3
go: upgraded github.com/kamalshkeir/ksbus v1.1.1 => v1.1.2
```

**Reply by kamalshkeir** [reply]
> > be aware , there is no more /name:str or others , only '/:name' is allowed
> 
> Unfortunately, right away from the update I face an exception. If you leave or will have only 1 route like ```sbus.App.POST("/:id", v1.ProcessSomething)``` you will have: 
> ```github.com/kamalshkeir/kmux.(*node).insertChild[line:579]:wildcard segment ':id' conflicts with existing children in path '/:id'```
> Doesn't matter the name ":id" or ":whatever" you will have this exception
> ```
> go: upgraded github.com/kamalshkeir/kmux v1.9.2 => v1.9.4
> go: upgraded github.com/kamalshkeir/korm v1.7.2 => v1.7.3
> go: upgraded github.com/kamalshkeir/ksbus v1.1.1 => v1.1.2
> ```
> 

this conflict should happen only when you have multiple routes for a given endpoint like /:id and /anything , i will check it

**Reply by manson** [reply]
> this conflict should happen only when you have multiple routes for a given endpoint like /:id and /anything , i will check it

Well... If I set 2 routes like: ``` /hello``` and ```/:id``` in this order, there should not rise any conflicts. If ```hello``` comes - it workes, if something else - ```/:id``` should work. But if I set first ```/:id``` and second ```/hello```, second would never rise because first one already intercepted it as valid.
But in this situation when I checked it to find where the problem is, I found that the single ```/:id``` route rises this exception (no other routes exist in app)

**Reply by kamalshkeir** [reply]
done @manson  ==> korm v1.4.7 , precise routes will arrive to the precise handler , others will match with the param or catchall like /:name or /*name, it took me longtime, i didn't want to loose the high performance, i bet it is faster than fiber and gin router ;)

**Reply by kamalshkeir** [reply]
Oh fuck, @manson , i miss one thing, big thing lol , all router Methods are now Titled not all capital, like app.GET => app.Get, i think it is prettier

**Reply by manson** [reply]
@kamalshkeir Well done! :-)
Looks like it works. Although I doubt that this is faster than fiber - as far as I remember, they use a special low level tuned network library written in C, some tweaks related to the memory allocations and consumption. But nevertheless it is great you think so :-) In any case I believe it is fast enough to not think of it the nearest 10000 concurrent users of an application, developed with Kago. :-)

**Reply by kamalshkeir** [reply]
> as far as I remember, they use a special low level tuned network library written in C, some tweaks related to the memory allocations and consumption

But this does not mean it is working very fast unless you benchmark, here are some benchmarks I have posted on LinkedIn https://www.linkedin.com/feed/update/urn:li:activity:7025770867661922305?utm_source=share&utm_medium=member_desktop, go with his model spawning a go routine for each request, very light, and so very fast, faster than rust and fasthttp which is what fiber is built on top of, also the new router is going to 44000 req/s and only 2 allocations on each request, also fasthttp not compatible with std library

**Reply by manson** [reply]
@kamalshkeir 
Meanwhile, could you please explain me, if possible, the next thing. I have used sse before in other projects, used regular EventSource  at javascript side and other client libraries. It keeps connection to a server very long time. It does not reconnect every 10 seconds as it does with Kago. I have replaced bus.App.SSE to  bus.App.Get and bind there the side sse-server just to check, and it still reconnects every about 10 seconds. I used Echo before, made the same thing, and connected client stands long while connection is broken (or at least several dozens of minutes, or even maybe an hour).  The problem is: as opposed to long polling, sse connections doesn't drop on every data income, it is a stream where data flows to a client uninterruptedly. So if data comes more or less often, this reconnection slows down the client data acceptance for about 2-4 seconds. This is not good. By the method of elimination, I came to the conclusion that you somehow control connections and not let them last long. Is that true? In any case, could consult me, how to make this behave like in other frameworks - that connection would last long without interrupts of itself?

UPD: Tested again - just pointed the same testing page against my other project (with Echo framework) with sse route, and it stands without interrupts.

**Reply by kamalshkeir** [reply]
@manson it's because of kmux.ReadTimeout and IdleTimeout and maybe WriteTimeout , try to play with these values until i understand how this is working

**Reply by manson** [reply]
@kamalshkeir 
Yea... I used this:
```
kmux.WriteTimeout = 5 * time.Minute
kmux.IdleTimeout = 5 * time.Minute
```
And it solved. Just making keepalive 1 minutes empty packets send and it lives well. Without keepalive packets some usefull packet (first) when long idle is lost. So, for now, I consider it as suites my needs.

**Reply by kamalshkeir** [reply]
@manson , some updates to SSE, here is an example i was playing with, working fine with the new updates, what i did is : instead of changing WriteTimeout and IdleTimeout globaly, i used a new go1.20 feature called `http.ResponseController` which can control these values for each request, so for SSE handlers this will be done automatically and set the headers, also it support param in path if the path doesn't conflict with another GET path .

```go
package main

import (
	"fmt"
	"net/http"

	"github.com/kamalshkeir/kmux"
)

func main() {
	app := kmux.New()
	app.Get("/", func(c *kmux.Context) {
		c.Html("index.html", nil)
	})
	app.LocalTemplates("temps")

	msgChan := make(chan string)
	app.Sse("/test/:param", func(c *kmux.Context) {
		fmt.Println(c.Param("param"))
		notify := c.ResponseWriter.(http.CloseNotifier).CloseNotify()
		for {
			select {
			case v := <-msgChan:
				err := c.Stream(v)
				if err != nil {
					return
				}
			case <-notify:
				return
			}
		}
	})

	// when someone hit this endpoint we publish to SSE using msgChan
	app.Get("/pub/sse", func(c *kmux.Context) {
		go func() {
			msgChan <- "hello stream"
		}()
		c.Text("sse")
	})

	app.Run(":9313")
}

```

```sh
go get -u github.com/kamalshkeir/kmux@v1.10.5
```

**Reply by manson** [reply]
@kamalshkeir 
I have checked new SSE. It works perfect for me. Way to go! I've switched back to this natural kago way to use SSE and satisfied. Thanks!
